Child map simplified

Arthur-Milchior opened the pull request:
`childMap` is not really easy to understand first. This hope to clarify what a Node is by using it for a single purpose instead of using it for two distinct purpose.

The second commit add comment and uses better variable name

david-allison-1 reviewed (changes requested):
We should remain consistent with the implementation from `libAnki`. 

The more we change things, the more difficult it gets to consolidate the interfaces between Decks.java and the V16 implementation of the class.

Arthur-Milchior commented:
I see you requested change. Can you please be more specific?

All of the methods I modified have disappeared from libanki since quite some time.

In c3b2b8625e048307613b876a7574c1a29173c65c I added in libanki the type:
```python
    childMapNode = Dict[int, Any]
    # Change to Dict[int, "DeckManager.childMapNode"] when MyPy allow recursive type
```
And I made the same mistake I made later in AnkiDroid, I used this type even in place where it didn't make sens. I'd correct it today if it were possible, however since the method is deleted, I can't correct it anymore. Does not mean it was not a mistake in the first place.

Also, we were never consistent. Since Java do not have a limitation on recursive type the same way MyPy has, we used recursive type where it made sens. (Anyway, being consistent with python typing makes little sens in the first place, given the huge use of Any in libanki)


If you want we keep `pid` instead of `parent_id`, I'd reverse this change. Same with introducing an intermediate variable. Not sure that it actually help in any way, since it's not related to the interface.

Arthur-Milchior commented:
I see you requested change. Can you please be more specific?

All of the methods I modified have disappeared from libanki since quite some time.

In c3b2b8625e048307613b876a7574c1a29173c65c I added in libanki the type:
```python
    childMapNode = Dict[int, Any]
    # Change to Dict[int, "DeckManager.childMapNode"] when MyPy allow recursive type
```
And I made the same mistake I made later in AnkiDroid, I used this type even in place where it didn't make sens. I'd correct it today if it were possible, however since the method is deleted, I can't correct it anymore. Does not mean it was not a mistake in the first place.

Also, we were never consistent. Since Java do not have a limitation on recursive type the same way MyPy has, we used recursive type where it made sens. (Anyway, being consistent with python typing makes little sens in the first place, given the huge use of Any in libanki)


If you want we keep `pid` instead of `parent_id`, I'd reverse this change. Same with introducing an intermediate variable. Not sure that it actually help in any way, since it's not related to the interface.

david-allison-1 commented:
After a few pull requests are merged, I'm going to need to combine the implementations of `DecksV16.kt` and `Decks.java` 

(see: https://github.com/ankidroid/Anki-Android/pull/9391/files).

The more that we change the interface away from how the Python was, or currently is, the more difficult this becomes.

The methods exist in the version of `libAnki` that we're currently compatible with: https://github.com/ankitects/anki/blob/c4db4bd2913234d077aa289543da6405a62f53dc/pylib/anki/decks.py

----

> Also, we were never consistent. Since Java do not have a limitation on recursive type the same way MyPy has, we used recursive type where it made sens. (Anyway, being consistent with python typing makes little sens in the first place, given the huge use of Any in libanki)

As a rule, we shouldn't be making changes to `libAnki` that aren't in upstream, as it makes maintenance and keeping compatibility with upstream much more difficult. We should aim to be as consistent as possible, focusing on maintainability, with readability of the code and efficiency as a secondary concern to this. It's easier to maintain code `Dict<str, Any>` that matches the Python than it is to deal with the mental overhead of understanding what a `Dict<str, Node>` entails.

Kotlin also forbids recursive definitions via `typealias`, so this matches the Python.

I'd accept a change in the style of #9391 to make the interfaces compatible, either from the `Decks` side, or the `DecksV16` side.

Note that this may conflict with https://github.com/ankidroid/Anki-Android/pull/9392
